### PR TITLE
Check round ids

### PR DESCRIPTION
**Pull request by eduard-cl**
Add same `epochAndRound` check to the contract and `_checkRounds()` to the tests.

### CODE REVIEW

File: contracts/src/v0.8/feeds/DualAggregator.sol [hunk 1]
```diff
@@ -826,6 +826,10 @@ contract DualAggregator is OCR2Abstract, OwnerIsCreator, AggregatorV2V3Interface
     // Report epoch and round
     uint40 epochAndRound = uint40(uint256(reportContext[1]));
 
+    if (epochAndRound == s_hotVars.latestEpochAndRound) {
+      return;
```
Comment (Oozyx) [inline contracts/src/v0.8/feeds/DualAggregator.sol]: Need to call _payTransmitter before returning

**Comment by Oozyx** [comment]
Only checking the epochAndRound doesn't work. We can end up with a situation where the transmitter constantly resends the same transmission and it won't revert.

File: contracts/src/v0.8/feeds/DualAggregator.sol [hunk 2]
```diff
@@ -816,25 +816,24 @@ contract DualAggregator is OCR2Abstract, OwnerIsCreator, AggregatorV2V3Interface
...
-    }
+    if (epochAndRound != s_hotVars.latestEpochAndRound || !s_primaryLocked) {
+      if (epochAndRound <= s_hotVars.latestEpochAndRound) {
+        revert StaleReport();
+      }
 
-    // Verify signatures attached to report
-    _verifySignatures(reportContext, report, rs, ss, rawVs);
+      // Verify signatures attached to report
+      _verifySignatures(reportContext, report, rs, ss, rawVs);
```
Comment (Oozyx) [inline contracts/src/v0.8/feeds/DualAggregator.sol]: Just realized we should always verify signatures. This should be placed under _validateReport

File: contracts/src/v0.8/feeds/DualAggregator.sol [hunk 3]
```diff
@@ -816,25 +816,24 @@ contract DualAggregator is OCR2Abstract, OwnerIsCreator, AggregatorV2V3Interface
         _payTransmitter(s_hotVars, report_.juelsPerFeeCoin, uint32(initialGas), msg.sender);
```
Comment (Oozyx) [inline contracts/src/v0.8/feeds/DualAggregator.sol:824]: It's possible the existing report is older than the current secondary round id.

File: contracts/src/v0.8/feeds/DualAggregator.sol [hunk 4]
```diff
@@ -816,25 +816,24 @@ contract DualAggregator is OCR2Abstract, OwnerIsCreator, AggregatorV2V3Interface
...
+      // Verify signatures attached to report
+      _verifySignatures(reportContext, report, rs, ss, rawVs);
 
-    int192 juelsPerFeeCoin = _report(s_hotVars, reportContext[0], epochAndRound, report_, isSecondary);
-    _payTransmitter(s_hotVars, juelsPerFeeCoin, uint32(initialGas), msg.sender);
+      _report(s_hotVars, reportContext[0], epochAndRound, report_, isSecondary);
+    }
+
+    // Lock/unlock the primary feed and pay the transmitters
+    s_primaryLocked = isSecondary;
```
Comment (alejoberardino) [inline contracts/src/v0.8/feeds/DualAggregator.sol]: This is not easy to understand, as we're only locking for a single block. Let's change this variable to something more direct, saying that the last update was from the secondary feed, and then communicate the implications of this on the reading side

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 2 above)
Comment (eduard-cl) [reply contracts/src/v0.8/feeds/DualAggregator.sol]: Why? To optimize gas costs it's better if we only verify the signatures one time per report.

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 3 above)
Comment (eduard-cl) [reply contracts/src/v0.8/feeds/DualAggregator.sol:824]: If an older report arrives it'll be `epochAndRound < s_hotVars.latestEpochAndRound` and revert with a `StaleReport()` error.

Comment (Oozyx) [reply contracts/src/v0.8/feeds/DualAggregator.sol:824]: I think you corrected it with lines 813-816

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 2 above)
Comment (Oozyx) [reply contracts/src/v0.8/feeds/DualAggregator.sol]: It's an attack vector. You can provide existing reports to the secondary feed that were not signed.

File: contracts/src/v0.8/feeds/DualAggregator.sol [hunk 5]
```diff
@@ -810,31 +810,36 @@ contract DualAggregator is OCR2Abstract, OwnerIsCreator, AggregatorV2V3Interface
...
     // Report epoch and round
     uint40 epochAndRound = uint40(uint256(reportContext[1]));
 
-    if (epochAndRound < s_hotVars.latestEpochAndRound) {
-      revert StaleReport();
-    }
+    if (epochAndRound != s_hotVars.latestEpochAndRound || !s_latestSecondary) {
+      if (epochAndRound <= s_hotVars.latestEpochAndRound) {
+        revert StaleReport();
+      }
```
Comment (Oozyx) [inline contracts/src/v0.8/feeds/DualAggregator.sol]: I'm finding these conditions hard to follow. Can you include some comments explaining what different edge cases we are validating for?

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 4 above)
Comment (eduard-cl) [reply contracts/src/v0.8/feeds/DualAggregator.sol]: That's true, renamed it to `s_latestSecondary`

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 2 above)
Comment (eduard-cl) [reply contracts/src/v0.8/feeds/DualAggregator.sol]: Yes, but the existing reports have to have the same timestamp and answer than the one already validated. I thought the biggest benefit of this implementation was to not have to double check the signatures if a report has previously arrived by the other aggregator path.

File: contracts/src/v0.8/feeds/DualAggregator.sol [hunk 6]
```diff
@@ -810,31 +810,36 @@ contract DualAggregator is OCR2Abstract, OwnerIsCreator, AggregatorV2V3Interface
...
+
         s_hotVars.latestSecondaryRoundId = roundId;
         emit SecondaryRoundIdUpdated(roundId);
 
         _payTransmitter(s_hotVars, report_.juelsPerFeeCoin, uint32(initialGas), msg.sender);
         return;
       }
-      // In case the report doesn't exist, lock the primary feed
-      s_primaryLocked = true;
-    } else {
```
Comment (Oozyx) [inline contracts/src/v0.8/feeds/DualAggregator.sol:821]: suggestion: I think we could keep this else and use it for the case where the primary feed deals with receiving an existing, but latest report.

`if (epochAndRound == s_hotVars.latestEpochAndRound && s_primaryLocked))`

I see this if, else block as where the logic is evaluated for all edge cases where an existing report is sent to either feed. Early returns would happen here as well.

The rest of the function would be to handle and record new reports.

This is very subjective so feel free to disagree.

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 2 above)
Comment (eduard-cl) [reply contracts/src/v0.8/feeds/DualAggregator.sol]: Moved it under `_validateReport()`, so we already have it.

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 5 above)
Comment (eduard-cl) [reply contracts/src/v0.8/feeds/DualAggregator.sol]: Added some comments, it's true it's a bit hard to understand, because I inverted some conditionals to not have duplicated code. To help you understand the code, it's the same as this:

```
// In case the report is the same as the latest and it arrived throw the secondary feed,
// we have to validate it without transmitting a duplicated report
if (epochAndRound == s_hotVars.latestEpochAndRound && s_latestSecondary) {
  s_latestSecondary = isSecondary;
  _payTransmitter(s_hotVars, report_.juelsPerFeeCoin, uint32(initialGas), msg.sender);
  return;
}

// revert because it's older or already arrived throw the primary feed too (if not we would 
// have match the previous check)
if (epochAndRound <= s_hotVars.latestEpochAndRound) {
  revert StaleReport();
}

_report(s_hotVars, reportContext[0], epochAndRound, report_, isSecondary);

// Store if the latest report was secondary or not
s_latestSecondary = isSecondary;
_payTransmitter(s_hotVars, report_.juelsPerFeeCoin, uint32(initialGas), msg.sender);
```

The only difference is that I inverted the first conditional to not have this lines duplicated:
```
s_latestSecondary = isSecondary;
_payTransmitter(s_hotVars, report_.juelsPerFeeCoin, uint32(initialGas), msg.sender);
```

File: contracts/src/v0.8/feeds/DualAggregator.sol (hunk 6 above)
Comment (eduard-cl) [reply contracts/src/v0.8/feeds/DualAggregator.sol:821]: Could you develop more this idea? Not sure I'm following how it should be implemented.

**Review by Oozyx — APPROVED** [review]

File: contracts/src/v0.8/feeds/DualAggregator.sol [hunk 7]
```diff
@@ -804,37 +804,46 @@ contract DualAggregator is OCR2Abstract, OwnerIsCreator, AggregatorV2V3Interface
     // Validate the report data
     _validateReport(reportContext, report, rs, ss);
 
+    // Verify signatures attached to report
+    _verifySignatures(reportContext, report, rs, ss, rawVs);
```
Comment (pszal) [inline contracts/src/v0.8/feeds/DualAggregator.sol:808]: Why it got moved? I have to be missing something, but isn't the main gas saving due to skipping this step if the report is already recorded?
